### PR TITLE
Security review PR - Collection of all "team project" feature related changes

### PR DESCRIPTION
**Pull request by pieterlukasse**
Main Jira Ticket: [VADC-618](https://ctds-planx.atlassian.net/browse/VADC-618)

### New Features
- Addition of "team project" access control to relevant cohort-middleware endpoints

[VADC-618]: https://ctds-planx.atlassian.net/browse/VADC-618?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: utils/parsing.go [hunk 1]
```diff
@@ -266,3 +275,28 @@ func ParseSourceIdAndCohortIdAndVariablesAsSingleList(c *gin.Context) (int, int,
...
+	for _, num := range input {
+		if !uniqueMap[num] {
+			uniqueMap[num] = true
+			uniqueList = append(uniqueList, num)
+		}
+	}
+	return uniqueList
+}
+
+func GetUniqueCohortDefinitionIdsListFromRequest(cohortDefinitionIds []int, filterCohortPairs []CustomDichotomousVariableDef) []int {
```
Comment (pieterlukasse) [inline utils/parsing.go:292]: TODO - add comment to method to clarify its purpose

File: models/cohortdefinition.go [hunk 2]
```diff
@@ -67,7 +69,36 @@ func (h CohortDefinition) GetAllCohortDefinitions() ([]*CohortDefinition, error)
...
-func (h CohortDefinition) GetAllCohortDefinitionsAndStatsOrderBySizeDesc(sourceId int) ([]*CohortDefinitionStats, error) {
+// Returns any "team project" entries that are matched to _each and every one_ of the
+// cohort definition ids found in uniqueCohortDefinitionIdsList.
+func (h CohortDefinition) GetTeamProjectsThatMatchAllCohortDefinitionIds(uniqueCohortDefinitionIdsList []int) ([]string, error) {
+
+	db2 := db.GetAtlasDB().Db
+	var teamProjects []string
+	// Find any roles that are paired to each and every one of the cohort_definition_id values.
+	// Roles that ony match part of the values are filtered out by the having(count) clause:
+	query := db2.Table(db.GetAtlasDB().Schema+".cohort_definition_sec_role").
```
Comment (k-burt-uch) [inline models/cohortdefinition.go:80]: Stating for my own understanding/clarity: 
https://github.com/uc-cdis/cohort-middleware/pull/89/files?diff=unified&w=1#diff-8d6e2a5018c8db98f284089e2369dd804849c0488d7166974d49992d28bad403R79
So based on the distinct here in this view, then this example::

```
sec_role_name | cohort_definition_id
--------------------------------------
foo                      | 8675309
foo                      | 8675309
```
should not be possible, right? Therefore theres no way for the count in this Having clause to double count the same cohort_definition_id such that it matches the length of uniqueCohortDefinitionList without every cohort_definition_id being present in the uniqueCohortDefinitionIdsList.

File: server/router.go [hunk 3]
```diff
@@ -26,14 +28,15 @@ func NewRouter() *gin.Engine {
...
-		authorized.GET("/cohortdefinition/by-name/:name", cohortdefinitions.RetriveByName)
-		authorized.GET("/cohortdefinitions", cohortdefinitions.RetriveAll)
-		authorized.GET("/cohortdefinition-stats/by-source-id/:sourceid", cohortdefinitions.RetriveStatsBySourceId)
+
+		authorized.GET("/cohortdefinition-stats/by-source-id/:sourceid/by-team-project", cohortdefinitions.RetriveStatsBySourceIdAndTeamProject)
 
 		// concept endpoints:
-		concepts := controllers.NewConceptController(*new(models.Concept), *new(models.CohortDefinition))
+		concepts := controllers.NewConceptController(*new(models.Concept), *new(models.CohortDefinition),
+			middlewares.NewTeamProjectAuthz(*new(models.CohortDefinition), &http.Client{}))
```
Comment (k-burt-uch) [inline server/router.go:39]: for my own understanding:

The only time we need to check the team authz is when interacting with cohorts, right? The concepts themselves don't need team authz unless they're touching the cohorts?

File: models/cohortdefinition.go (hunk 2 above)
Comment (pieterlukasse) [reply models/cohortdefinition.go:80]: The count for above would return 1. The goal of this check is to ensure that in the scenario below:
```
sec_role_name | cohort_definition_id  
--------------------------------------  
foo                      | 8675308  
foo                      | 8675309 
```
and `uniqueCohortDefinitionIdsList=[8675308, 8675310]` 
the `foo` team is NOT returned. The `Where ... in ...` clause will match,
but the `Having` will not in this case.

Comment (pieterlukasse) [reply models/cohortdefinition.go:80]: This is a good scenario to add to the unit tests...

File: server/router.go (hunk 3 above)
Comment (pieterlukasse) [reply server/router.go:39]: correct.

File: models/cohortdefinition.go (hunk 2 above)
Comment (pieterlukasse) [reply models/cohortdefinition.go:80]: Test added.

**Review by k-burt-uch — APPROVED** [review]

**Comment by pieterlukasse** [comment]
thanks for the approval @k-burt-uch . I'll move just the new commits to a separate PR to `master`. Closing this one and keeping it for the record.

File: utils/parsing.go (hunk 1 above)
Comment (pieterlukasse) [reply utils/parsing.go:301]: addressed in https://github.com/uc-cdis/cohort-middleware/pull/90

**Pull request closed without merging**